timemachine/cpp/src/local_md_potentials.cu

```diff
@@ -0,0 +1,191 @@
+    d_temp_storage_buffer_.reset(new DeviceBuffer<char>(temp_storage_bytes_));
+
+    curandErrchk(curandCreateGenerator(&cr_rng_, CURAND_RNG_PSEUDO_DEFAULT));
+};
+
+LocalMDPotentials::~LocalMDPotentials() { curandErrchk(curandDestroyGenerator(cr_rng_)); }
+
+// setup_from_idxs takes a set of idxs, a temperature and a seed to determine the free particles. Fix the local_idxs to length
+// one to ensure the same reference everytime, though the seed also handles the probabilities of selecting particles, and it is suggested
+// to provide a new seed at each step.
+void LocalMDPotentials::setup_from_idxs(
+    double *d_x_t,
+    double *d_box_t,
+    const std::vector<int> &local_idxs,
+    const double temperature,
+    const int seed,
+    const double radius,
+    const double k,
+    const cudaStream_t stream) {
+
+    curandErrchk(curandSetStream(cr_rng_, stream));
+    curandErrchk(curandSetPseudoRandomGeneratorSeed(cr_rng_, seed));
+    // Reset the generator offset to ensure same values for the same seed are produced
+    // Simply reseeding does NOT produce identical results
+    curandErrchk(curandSetGeneratorOffset(cr_rng_, 0));
+
+    // Set the array to all N, which indicates to ignore that idx
+    k_initialize_array<unsigned int><<<ceil_divide(N_, warp_size), warp_size, 0, stream>>>(N_, d_free_idxs_.data, N_);
+    gpuErrchk(cudaPeekAtLastError());
+
+    // Generate values between (0, 1.0]
+    curandErrchk(curandGenerateUniform(cr_rng_, probability_buffer_.data, round_up_even(N_)));
+
+    std::mt19937 rng;
+    rng.seed(seed);
+    std::uniform_int_distribution<unsigned int> random_dist(0, local_idxs.size() - 1);
+
+    unsigned int reference_idx = local_idxs[random_dist(rng)];
+
+    const double kBT = BOLTZ * temperature;
+    // Select all of the particles that will be free
+    k_log_probability_selection<float><<<ceil_divide(N_, warp_size), warp_size, 0, stream>>>(
+        N_, kBT, radius, k, reference_idx, d_x_t, d_box_t, probability_buffer_.data, d_free_idxs_.data);
+    gpuErrchk(cudaPeekAtLastError());
+
+    this->_setup_free_idxs_given_reference_idx(reference_idx, radius, k, stream);
+}
+
+void LocalMDPotentials::_setup_free_idxs_given_reference_idx(
+    const unsigned int reference_idx, const double radius, const double k, const cudaStream_t stream) {
+    const int tpb = warp_size;
+
+    LessThan select_op(N_);
+
+    // Partition the free idxs into the row idxs
+    gpuErrchk(cub::DevicePartition::If(
+        d_temp_storage_buffer_->data,
+        temp_storage_bytes_,
+        d_free_idxs_.data,
+        d_row_idxs_.data,
+        num_selected_buffer_.data,
+        N_,
+        select_op,
+        stream));
+
+    gpuErrchk(cudaMemcpyAsync(
+        p_num_selected_.data,
+        num_selected_buffer_.data,
+        1 * sizeof(*p_num_selected_.data),
+        cudaMemcpyDeviceToHost,
+        stream));
+    gpuErrchk(cudaStreamSynchronize(stream));
+
+    const int num_row_idxs = p_num_selected_.data[0];
```

Comment:
can you add a comment indicating what the `row_idxs` and `col_idxs` correspond to?

Reply:
(discussing offline: row_idxs: free_idxs but no reference, col_idxs: reference + everything else.)

Reply:
Added comment in https://github.com/proteneer/timemachine/pull/1004/commits/4e35e5e1a439746ae038507419311b924a0cd309